### PR TITLE
new generator `historyRoll`

### PR DESCRIPTION
**Pull request by ImpleLee**

### CODE REVIEW

**Comment by ImpleLee** [comment]
@aegistudio This is ready for review.

File: model/src/generator.cpp [hunk 1]
```diff
@@ -53,5 +53,34 @@ const tile* tilePermutator::generate() {
 	return result;
 }
 
+historyRoll::historyRoll(
+	const tile* tiles[], std::size_t numTiles,
```
Comment (aegistudio) [inline model/src/generator.cpp]: unused argument tiles here.

**Comment by aegistudio** [comment]
I don't think this piece of code will function, as long as the `tiles` field in `historyRoll` class is not initialized with concrete value.

File: model/src/generator.cpp [hunk 2]
```diff
@@ -53,5 +53,34 @@ const tile* tilePermutator::generate() {
...
+	history(initialHistory, initialHistory + historySize),
+	tiles(new const tile*[numTiles]), numTiles(numTiles),
+	retryTimes(retryTimes), historySize(historySize),
+	randomizer(seed) {}
+
+const tile* historyRoll::generate() {
+	std::uniform_int_distribution<std::size_t> sampler{0, numTiles - 1};
+	std::size_t result = sampler(randomizer);
+	for (std::size_t i = 0; i < retryTimes; ++i) {
+		// TODO: better way to find previously generated tiles
```
Comment (aegistudio) [inline model/src/generator.cpp]: It seems that a deque plus an array of length `numTiles` storing the reference count will do.

Comment (aegistudio) [reply model/src/generator.cpp]: When a tile is enqueued, the reference counter increments, and vice versa.

File: model/src/generator.cpp (hunk 1 above)
Comment (ImpleLee) [reply model/src/generator.cpp]: fixed in 72ea34b

File: model/src/generator.cpp (hunk 2 above)
Comment (ImpleLee) [reply model/src/generator.cpp]: applied in c76f353

File: model/generator.hpp [hunk 3]
```diff
@@ -12,6 +12,8 @@
  */
 #include "model/tile.hpp"
 #include <memory>
+#include <random>
```
Comment (aegistudio) [inline model/generator.hpp]: useless include

File: model/src/generator.cpp [hunk 4]
```diff
@@ -53,5 +53,40 @@ const tile* tilePermutator::generate() {
...
+}
+
+const tile* historyRoll::generate() {
+	std::uniform_int_distribution<std::size_t> sampler{0, numTiles - 1};
+	std::size_t result = sampler(*reinterpret_cast<randomizer*>(workData));
+	for (std::size_t i = 0; i < retryTimes && counts[result] > 0; ++i)
+		result = sampler(*reinterpret_cast<randomizer*>(workData));
+	history.push_back(result);
+	counts[result]++;
+	if (history.front() < numTiles)
```
Comment (aegistudio) [inline model/src/generator.cpp:85]: Why won't the element be popped immediately on line 85 after it is pushed into the deque on line 83?

File: model/generator.hpp (hunk 3 above)
Comment (ImpleLee) [reply model/generator.hpp]: fixed in a17f67c.

**Pull request merged**